components/layout/block.rs

```diff
@@ -1456,32 +1457,72 @@ impl BlockFlow {
         debug_assert!(self.formatting_context_type() != FormattingContextType::None);
 
         if !self.base.restyle_damage.intersects(REFLOW_OUT_OF_FLOW | REFLOW) {
             return
         }
 
-        let info = PlacementInfo {
-            size: self.fragment.border_box.size.convert(self.fragment.style.writing_mode,
-                                                        self.base.floats.writing_mode),
-            ceiling: self.base.position.start.b,
-            max_inline_size: MAX_AU,
-            kind: FloatKind::Left,
-        };
+        // We do this first to avoid recomputing our inline size when we propagate it.
+        self.base.restyle_damage.remove(REFLOW_OUT_OF_FLOW | REFLOW);
+        self.fragment.restyle_damage.remove(REFLOW_OUT_OF_FLOW | REFLOW);
 
         // Offset our position by whatever displacement is needed to not impact the floats.
-        let rect = self.base.floats.place_between_floats(&info);
-        self.base.position.start.i = self.base.position.start.i + rect.start.i;
+        // N.B. this call to available_rect computes the available space for the border box,
+        // so the area taken up by margins is not available.
+        let rect = self.base.floats.available_rect(
+            self.base.position.start.b,
+            self.fragment.border_box.size.block,
+            self.base.block_container_inline_size - self.fragment.margin.inline_start_end(),
+        );
+        match rect {
+            Some(rect) => {
+                self.base.position.start.i = max(
+                    rect.start.i,
+                    self.fragment.margin.inline_start,
+                );
+                let containing_block_inline_size = self.base.block_container_inline_size;
+                let max_inline_size = specified_or_none(
+                    self.fragment.style().max_inline_size(),
+                    containing_block_inline_size
+                ).unwrap_or(MAX_AU);
+                let min_inline_size = specified(
+                    self.fragment.style().min_inline_size(),
+                    containing_block_inline_size
+                );
+                let avail_inline_size = if rect.size.inline > max_inline_size {
+                    max_inline_size
+                } else if rect.size.inline < min_inline_size {
+                    min_inline_size
+                } else {
+                    rect.size.inline
+                };
+                self.base.position.size.inline = avail_inline_size
+                    + self.fragment.margin.inline_start_end();
+            }
+            None => {}
+        }
 
-        // TODO(pcwalton): If the inline-size of this flow is different from the size we estimated
-        // earlier, lay it out again.
+        // If float speculation failed, fixup our layout, and re-layout all the children.
+        if self.fragment.margin_box_inline_size() != self.base.position.size.inline {
```

Comment:
Could you add a `debug!()` log message here? For performance reasons, it's nice to know when we hit this case.